src/backend/drm/Renderer.cpp

```diff
@@ -846,22 +893,40 @@ CDRMRenderer::SBlitResult CDRMRenderer::blit(SP<IBuffer> from, SP<IBuffer> to, i
     // both from and to have the same AQ_ATTACHMENT_DRM_RENDERER_DATA.
     // Those buffers always come from different swapchains, so it's OK.
 
-    SGLTex fromTex;
+    SGLTex             fromTex;
+    auto               fromDma = from->dmabuf();
+    std::span<uint8_t> intermediateBuf;
     {
         auto attachment = from->attachments.get(AQ_ATTACHMENT_DRM_RENDERER_DATA);
         if (attachment) {
             TRACE(backend->log(AQ_LOG_TRACE, "EGL (blit): From attachment found"));
-            auto att = (CDRMRendererBufferAttachment*)attachment.get();
-            fromTex  = att->tex;
+            auto att        = (CDRMRendererBufferAttachment*)attachment.get();
+            fromTex         = att->tex;
+            intermediateBuf = att->intermediateBuf;
         }
 
-        if (!fromTex.image) {
+        if (!fromTex.image && intermediateBuf.empty()) {
             backend->log(AQ_LOG_DEBUG, "EGL (blit): No attachment in from, creating a new image");
             fromTex = glTex(from);
 
+            auto newAttachment = makeShared<CDRMRendererBufferAttachment>(self, from, nullptr, 0, 0, fromTex, std::vector<uint8_t>());
+            if (!fromTex.image && primaryRenderer) {
+                backend->log(AQ_LOG_DEBUG, "EGL (blit): Failed to create image from source buffer directly, allocating intermediate buffer");
+                static_assert(PIXEL_BUFFER_FORMAT == GL_RGBA); // If the pixel buffer format changes, the below size calculation probably needs to as well.
+                newAttachment->intermediateBuf.resize(fromDma.size.x * fromDma.size.y * 4);
+                intermediateBuf = newAttachment->intermediateBuf;
+                fromTex.target  = GL_TEXTURE_2D;
+                GLCALL(glGenTextures(1, &fromTex.texid));
+            }
+
             // should never remove anything, but JIC. We'll leak an EGLImage if this removes anything.
             from->attachments.removeByType(AQ_ATTACHMENT_DRM_RENDERER_DATA);
-            from->attachments.add(makeShared<CDRMRendererBufferAttachment>(self, from, nullptr, 0, 0, fromTex));
+            from->attachments.add(newAttachment);
+        }
+
+        if (!intermediateBuf.empty()) {
+            // Note: this might modify from's attachments
+            primaryRenderer->readBuffer(from, intermediateBuf);
```

Comment:
needs a null check. primaryRenderer is not guaranteed, and I remember some odd configurations where it is null

Reply:
primaryRenderer can definitely be null, but intermediateBuf should never be populated without a primary renderer. That said, absolutely better safe than sorry. I've added the null check